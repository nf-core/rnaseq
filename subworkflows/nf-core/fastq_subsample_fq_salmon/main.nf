//
// Sub-sample FastQ files and pseudo-align with Salmon
//      can be used to infer strandedness of library
//

include { SALMON_INDEX } from '../../../modules/nf-core/salmon/index/main'
include { FQ_SUBSAMPLE } from '../../../modules/nf-core/fq/subsample/main'
include { SALMON_QUANT } from '../../../modules/nf-core/salmon/quant/main'

workflow FASTQ_SUBSAMPLE_FQ_SALMON {
    take:
    ch_reads            // channel: [ val(meta), [ reads ] ]
    ch_genome_fasta     // channel: /path/to/genome.fasta
    ch_transcript_fasta // channel: /path/to/transcript.fasta
    ch_gtf              // channel: /path/to/genome.gtf
    ch_index            // channel: /path/to/salmon/index/
    make_index          // boolean: Whether to create salmon index before running salmon quant

    main:

    ch_versions = Channel.empty()

    //
    // Create Salmon index if required
    //
    if (make_index) {
        SALMON_INDEX.config.ext.args   = params.gencode ? '--gencode' : ''
        SALMON_INDEX.config.publishDir = [
            path: "${params.outdir}/genome/index",
            mode: params.publish_dir_mode,
            saveAs: { filename -> filename.equals('versions.yml') ? null : filename },
            enabled: params.save_reference
        ]
        ch_index = SALMON_INDEX ( ch_genome_fasta, ch_transcript_fasta ).index
        ch_versions = ch_versions.mix(SALMON_INDEX.out.versions)
    }

    //
    // Sub-sample FastQ files with fq
    //
    FQ_SUBSAMPLE.config.ext.args   = params.subsample_fq_args
    FQ_SUBSAMPLE.config.ext.prefix = { "${meta.id}.subsampled" }
    FQ_SUBSAMPLE.config.publishDir = [
        path: "${params.outdir}/sample_fastq/fastq",
        mode: params.publish_dir_mode,
        saveAs: { filename -> filename.equals('versions.yml') ? null : filename },
        enabled: false
    ]
    FQ_SUBSAMPLE ( ch_reads )
    ch_versions = ch_versions.mix(FQ_SUBSAMPLE.out.versions.first())

    //
    // Pseudo-alignment with Salmon
    //
    def lib_type = 'A'
    def alignment_mode = false
    SALMON_QUANT.config.ext.args   = params.subsample_salmon_args
    SALMON_QUANT.config.publishDir = [
        path: "${params.outdir}/sample_fastq/salmon",
        mode: params.publish_dir_mode,
        saveAs: { filename -> filename.equals('versions.yml') || filename.endsWith('_meta_info.json') ? null : filename },
        enabled: false
    ]
    SALMON_QUANT ( FQ_SUBSAMPLE.out.fastq, ch_index, ch_gtf, ch_transcript_fasta, alignment_mode, lib_type )
    ch_versions = ch_versions.mix(SALMON_QUANT.out.versions.first())

    emit:
    index     = ch_index                   // channel: [ index ]

    reads     = FQ_SUBSAMPLE.out.fastq     // channel: [ val(meta), fastq ]

    results   = SALMON_QUANT.out.results   // channel: [ val(meta), results_dir ]
    json_info = SALMON_QUANT.out.json_info // channel: [ val(meta), json_info

    versions  = ch_versions                // channel: [ versions.yml ]
}
