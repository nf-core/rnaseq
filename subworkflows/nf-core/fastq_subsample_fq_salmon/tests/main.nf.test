nextflow_workflow {

    name "Test Workflow FASTQ_SUBSAMPLE_FQ_SALMON"
    script "../main.nf"
    workflow "FASTQ_SUBSAMPLE_FQ_SALMON"
    config "./nextflow.config"
    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/fastq_subsample_fq_salmon"
    tag "fastq_subsample_fq_salmon"
    tag "salmon/index"
    tag "fq/subsample"
    tag "salmon/quant"

    test("homo_sapiens paired-end [fastq]") {

        setup {
            run("SALMON_INDEX") {
                script "../../../../modules/nf-core/salmon/index/main.nf"
                process {
                """
                input[0] = Channel.of(file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true))        // genome_fasta
                input[1] = Channel.of(file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/transcriptome.fasta', checkIfExists: true)) // transcriptome_fasta
                """
                }
            }
        }

        when {
            workflow {
                """
                make_index = false

                input[0] = Channel.of([
                    [ id:'test', single_end:false ], // meta map
                    [ file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/fastq/test_1.fastq.gz', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/fastq/test_2.fastq.gz', checkIfExists: true) ]
                ])
                input[1] = Channel.of(file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true))        // genome_fasta
                input[2] = Channel.of(file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/transcriptome.fasta', checkIfExists: true)) // transcriptome_fasta
                input[3] = Channel.of(file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.gtf', checkIfExists: true))          // genome_gtf
                input[4] = SALMON_INDEX.out.index
                input[5] = make_index
                """
            }
        }

        then {
            def readlines1 = path(workflow.out.reads[0][1][0]).linesGzip
            def readlines2 = path(workflow.out.reads[0][1][1]).linesGzip
            assertAll(
                { assert workflow.success },
                { assert snapshot(readlines1[0..5]).match("test_reads_1_lines") },
                { assert snapshot(readlines1.size()).match("test_reads_1_size") },
                { assert snapshot(readlines2[0..5]).match("test_reads_2_lines") },
                { assert snapshot(readlines2.size()).match("test_reads_2_size") },
                { assert snapshot(workflow.out.versions).match("versions") },

                { assert workflow.out.index },
                { assert workflow.out.results },
                { assert workflow.out.json_info }
            )
        }
    }
}
