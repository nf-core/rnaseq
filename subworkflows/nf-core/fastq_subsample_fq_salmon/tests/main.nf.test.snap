{
    "test_reads_1_size": {
        "content": [
            1066944
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-01-11T10:33:54.747992124"
    },
    "versions": {
        "content": [
            [
                "versions.yml:md5,12c0d1f67c2afb97470ae0974e5e01bb",
                "versions.yml:md5,885fde9e7beac002b3a17b66b92db4bd"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2023-11-26T16:41:10.396971682"
    },
    "test_reads_1_lines": {
        "content": [
            [
                "@normal#21#998579#1/1",
                "CCTTCTCCCTGCTGGGGTTGCTTGTCAGTAGCGGGCAAGGTAGGAGTGTGGCGCTTTATTGCATTTACTTTCCCTCCCCCTTCCCCCCGGCCAAGAGAGG",
                "+",
                "102302000331;3333;23133320233330*33/233333333333333/313232333/3;3;3/333000;11/00;;01//103*1032323233",
                "@normal#21#998572#2/1",
                "CTCCTCTCCTTCTACCTGCTGGGGTTGCTTGTCAGTAGCGGGCAAGGTCGGAGTGTTGCGCTTTATTGCATTTACTTTCCCTCCCCCTTCCACCCGGCCA"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-01-11T10:33:54.730250665"
    },
    "test_reads_2_lines": {
        "content": [
            [
                "@normal#21#998579#1/2",
                "AAAAAAAAAGAAGAAGCAGAAGCTGTTTCCCTGGATATCCTGCTCACCGATTCCCCTCTCCAATTCTGTATTTTCCCTTCTCTTATTTAAGGGTCTCCAC",
                "+",
                "023333233332333310333302333211/3333;0300;*/;000/32;201003031/22;21333032;;11/23030322;2332333313/030",
                "@normal#21#998572#2/2",
                "TTCCCCTCTCCAATTGAGTATTTTCCCTTCTCTTATTTAAGGGTCTCCACACAAACAGATACAATTTTAGGGACAGCTAGGAGAAAGAACGAAAATAATAA"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-01-11T10:33:54.756723613"
    },
    "test_reads_2_size": {
        "content": [
            1066944
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-01-11T10:33:54.763399473"
    }
}